# Loading data from Digital Earth Africa <img align="right" src="../Supplementary_data/DE_Africa_Logo_Stacked_RGB_small.jpg">

* **Products used:** 
[ga_ls8c_gm_2_annual](https://explorer.digitalearth.africa/ga_ls8c_gm_2_annual), 
[ga_ls8c_wofs_2_annual_summary](https://explorer.digitalearth.africa/ga_ls8c_wofs_2_annual_summary)

* **Prerequisites:** Users of this notebook should have a basic understanding of:
    * How to run a [Jupyter notebook](01_Jupyter_notebooks.ipynb)
    * Inspecting available [DE Africa products and measurements](02_Products_and_measurements.ipynb)

## Background
Loading data from the [Digital Earth Africa (DE Africa)](https://www.digitalearthafrica.org/) instance of the [Open Data Cube](https://www.opendatacube.org/) requires the construction of a data query that specifies the what, where, and when of the data request.
Each query returns a [multi-dimensional xarray object](http://xarray.pydata.org/en/stable/) containing the contents of your query.
It is essential to understand the `xarray` data structures as they are fundamental to the structure of data loaded from the datacube.
Manipulations, transformations and visualisation of `xarray` objects provide datacube users with the ability to explore and analyse DEA datasets, as well as pose and answer scientific questions.

## Description
This notebook will introduce how to load data from the Digital Earth Africa datacube through the construction of a query and use of the `dc.load()` function.
Topics covered include:

* Loading data using `dc.load()`
* Interpreting the resulting `xarray.Dataset` object
    * Inspecting an individual `xarray.DataArray`
* Customising parameters passed to the `dc.load()` function
    * Loading specific measurements
    * Loading data for coordinates in a custom coordinate reference system (CRS)
    * Projecting data to a new CRS and spatial resolution 
    * Specifying a specific spatial resampling method
* Loading data using a reusable dictionary query
* Loading matching data from multiple products using `like`
* Adding a progress bar to the data load

***

## Getting started
To run this introduction to loading data from DE Africa, run all the cells in the notebook starting with the "Load packages" cell. For help with running notebook cells, refer back to the [Jupyter Notebooks notebook](01_Jupyter_notebooks.ipynb).

### Load packages
First we need to load the `datacube` package.
This will allow us to query the datacube database and load some data. 
The `with_ui_cbk` function from `odc.ui` will allow us to show a progress bar when loading large amounts of data.

In [1]:
import datacube
from odc.ui import with_ui_cbk

### Connect to the datacube
We then need to connect to the datacube database.
We will then be able to use the `dc` datacube object to load data.
The `app` parameter is a unique name used to identify the notebook that does not have any effect on the analysis.

In [2]:
dc = datacube.Datacube(app="03_Loading_data")

## Loading data using `dc.load()`

Loading data from the datacube uses the [dc.load()](https://datacube-core.readthedocs.io/en/latest/dev/api/generate/datacube.Datacube.load.html) function.

The function requires the following minimum arguments:

* `product`: A specific product to load (to revise DE Africa products, see the [Products and measurements](02_Products_and_measurements.ipynb) notebook).
* `x`: Defines the spatial region in the *x* dimension. By default, the *x* and *y* arguments accept queries in a geographical co-ordinate system WGS84, identified by the EPSG code *4326*.
* `y`: Defines the spatial region in the *y* dimension. The dimensions ``longitude``/``latitude`` and ``x``/``y`` can be used interchangeably.
* `time`: Defines the temporal extent. The time dimension can be specified using a tuple of datetime objects or strings in the "YYYY", "YYYY-MM" or "YYYY-MM-DD" format. 

Let's run a query to load 2018 data from the [Landsat 8 annual geomedian product](https://explorer.digitalearth.africa/ga_ls8c_gm_2_annual)) for part of Nxai Pan National park in Botswana. 
For this example, we can use the following parameters:

* `product`: `ga_ls8c_gm_2_annual`
* `x`: `(24.60, 24.80)`
* `y`: `(-20.05, -20.25)`
* `time`: `("2018-01-01", "2018-12-31")`

Run the following cell to load all datasets from the `ga_ls8c_gm_2_annual` product that match this spatial and temporal extent:

In [3]:
ds = dc.load(product="ga_ls8c_gm_2_annual",
             x=(24.60, 24.80),
             y=(-20.05, -20.25),
             time=("2018-01-01", "2018-12-31"))

print(ds)

<xarray.Dataset>
Dimensions:  (time: 1, x: 644, y: 801)
Coordinates:
  * time     (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y        (y) float64 -2.507e+06 -2.507e+06 ... -2.531e+06 -2.531e+06
  * x        (x) float64 2.374e+06 2.374e+06 2.374e+06 ... 2.393e+06 2.393e+06
Data variables:
    red      (time, y, x) uint16 11808 11667 11901 11246 ... 11944 11948 11830
    green    (time, y, x) uint16 10801 10747 11000 10543 ... 11075 11068 10990
    blue     (time, y, x) uint16 9639 9581 9765 9467 ... 9892 9916 9926 9869
    nir      (time, y, x) uint16 15688 15618 15904 15293 ... 16541 16543 16316
    swir_1   (time, y, x) uint16 17885 17613 17969 16992 ... 18147 18170 17955
    swir_2   (time, y, x) uint16 14762 14453 14678 13979 ... 14850 14875 14693
Attributes:
    crs:      EPSG:6933


### Interpreting the resulting `xarray.Dataset`
The variable `ds` has returned an `xarray.Dataset` containing all data that matched the spatial and temporal query parameters inputted into `dc.load`.

*Dimensions* 

* Identifies the number of timesteps returned in the search (`time: 1`) as well as the number of pixels in the `x` and `y` directions of the data query.

*Coordinates* 

* `time` identifies the date attributed to each returned timestep.
* `x` and `y` are the coordinates for each pixel within the spatial bounds of your query.

*Data variables*

* These are the measurements available for the nominated product. 
For every date (`time`) returned by the query, the measured value at each pixel (`y`, `x`) is returned as an array for each measurement.
Each data variable is itself an `xarray.DataArray` object ([see below](#Inspecting-an-individual-xarray.DataArray)). 

*Attributes*

* `crs` identifies the coordinate reference system (CRS) of the loaded data. 

### Inspecting an individual `xarray.DataArray`
The `xarray.Dataset` we loaded above is itself a collection of individual `xarray.DataArray` objects that hold the actual data for each data variable/measurement. 
For example, all measurements listed under _Data variables_ above (e.g. `blue`, `green`, `red`, `nir`, `swir1`, `swir2`) are `xarray.DataArray` objects.

We can inspect the data in these `xarray.DataArray` objects using either of the following syntaxes:
```
ds["measurement_name"]
```
or:
```
ds.measurement_name
```

Being able to access data from individual data variables/measurements allows us to manipulate and analyse data from individual satellite bands or specific layers in a dataset. 
For example, we can access data from the near infra-red satellite band (i.e. `nir`):

In [4]:
print(ds.nir)

<xarray.DataArray 'nir' (time: 1, y: 801, x: 644)>
array([[[15688, 15618, ..., 15901, 15985],
        [16288, 16464, ..., 16092, 16084],
        ...,
        [17577, 17600, ..., 16491, 16312],
        [17659, 17691, ..., 16543, 16316]]], dtype=uint16)
Coordinates:
  * time     (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y        (y) float64 -2.507e+06 -2.507e+06 ... -2.531e+06 -2.531e+06
  * x        (x) float64 2.374e+06 2.374e+06 2.374e+06 ... 2.393e+06 2.393e+06
Attributes:
    units:    1
    nodata:   0
    crs:      EPSG:6933


Note that the object header informs us that it is an `xarray.DataArray` containing data for the `nir` satellite band. 

Like an `xarray.Dataset`, the array also includes information about the data's **dimensions** (i.e. `(time: 1, y: 801, x: 644)`), **coordinates** and **attributes**.
This particular data variable/measurement contains some additional information that is specific to the `nir` band, including details of array's nodata value (i.e. `nodata: -999`).

> **Note**: For a more in-depth introduction to `xarray` data structures, refer to the [official xarray documentation](http://xarray.pydata.org/en/stable/data-structures.html)

## Customising the `dc.load()` function

The `dc.load()` function can be tailored to refine a query.

Customisation options include:

* `measurements:` This argument is used to provide a list of measurement names to load, as listed in `dc.list_measurements()`. 
For satellite datasets, measurements contain data for each individual satellite band (e.g. near infrared). 
If not provided, all measurements for the product will be returned.
* `crs:` The coordinate reference system (CRS) of the query's `x` and `y` coordinates is assumed to be `WGS84`/`EPSG:4326` unless the `crs` field is supplied, even if the stored data is in another projection or the `output_crs` is specified. 
The `crs` parameter is required if your query's coordinates are in any other CRS.
* `group_by:` Satellite datasets based around scenes can have multiple observations per day with slightly different time stamps as the satellite collects data along its path.
These observations can be combined by reducing the `time` dimension to the day level using `group_by=solar_day`.
* `output_crs` and `resolution`: To reproject or change the resolution the data, supply the `output_crs` and `resolution` fields.    
* `resampling`: This argument allows you to specify a custom spatial resampling method to use when data is reprojected into a different CRS. 

Example syntax on the use of these options follows in the cells below.

> For help or more customisation options, run `help(dc.load)` in an empty cell or visit the function's [documentation page](https://datacube-core.readthedocs.io/en/latest/dev/api/generate/datacube.Datacube.load.html)


### Specifying measurements
By default, `dc.load()` will load *all* measurements in a product.

To load data from the `red`, `green` and `blue` satellite bands only, we can add `measurements=["red", "green", "blue"]` to our query:

In [5]:
# Note the optional inclusion of the measurements list
ds_rgb = dc.load(product="ga_ls8c_gm_2_annual",
                 measurements=["red", "green", "blue"],
                 x=(24.60, 24.80),
                 y=(-20.05, -20.25),
                 time=("2018-01-01", "2018-12-31"))

print(ds_rgb)

<xarray.Dataset>
Dimensions:  (time: 1, x: 644, y: 801)
Coordinates:
  * time     (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y        (y) float64 -2.507e+06 -2.507e+06 ... -2.531e+06 -2.531e+06
  * x        (x) float64 2.374e+06 2.374e+06 2.374e+06 ... 2.393e+06 2.393e+06
Data variables:
    red      (time, y, x) uint16 11808 11667 11901 11246 ... 11944 11948 11830
    green    (time, y, x) uint16 10801 10747 11000 10543 ... 11075 11068 10990
    blue     (time, y, x) uint16 9639 9581 9765 9467 ... 9892 9916 9926 9869
Attributes:
    crs:      EPSG:6933


Note that the *Data variables* component of the `xarray.Dataset` now includes only the measurements specified in the query (i.e. the `red`, `green` and `blue` satellite bands).

### Loading data for coordinates in any CRS
By default, `dc.load()` assumes that your query `x` and `y` coordinates are provided in degrees in the `WGS84/EPSG:4326` CRS.
If your coordinates are in a different coordinate system, you need to specify this using the `crs` parameter.

In the example below, we load data for a set of `x` and `y` coordinates defined in Africa Albers Equal Area Conic (`EPSG:102022`), and ensure that the `dc.load()` function accounts for this by including `crs="EPSG:102022"`:


In [6]:
# Note the new `x` and `y` coordinates and `crs` parameter
ds_custom_crs = dc.load(product="ga_ls8c_gm_2_annual",
                        time=("2018-01-01", "2018-12-31"),
                        x=(-34340, -12047),
                        y=(-2345844, -2363768),
                        crs="EPSG:102022")

print(ds_custom_crs)

<xarray.Dataset>
Dimensions:  (time: 1, x: 699, y: 637)
Coordinates:
  * time     (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y        (y) float64 -2.508e+06 -2.508e+06 ... -2.527e+06 -2.527e+06
  * x        (x) float64 2.38e+06 2.38e+06 2.38e+06 ... 2.401e+06 2.401e+06
Data variables:
    red      (time, y, x) uint16 12229 12156 12054 12021 ... 11815 11769 11769
    green    (time, y, x) uint16 11289 11224 11163 11176 ... 11097 11059 11127
    blue     (time, y, x) uint16 9933 9885 9821 9812 ... 10195 10094 10072 10108
    nir      (time, y, x) uint16 16417 16355 16326 16428 ... 16749 16791 16851
    swir_1   (time, y, x) uint16 18488 18296 18099 18082 ... 18835 18840 18841
    swir_2   (time, y, x) uint16 15244 15003 14822 14860 ... 15649 15670 15724
Attributes:
    crs:      EPSG:6933


### CRS reprojection
Certain applications may require that you output your data into a specific CRS.
You can reproject your output data by specifying the new `output_crs` and identifying the `resolution` required.

In this example, we will reproject our data to a new CRS (UTM Zone 34S, `EPSG:32734`) and resolution (250 x 250 m). Note that for most CRSs, the first resolution value is negative (e.g. `(-250, 250)`):

In [7]:
ds_reprojected = dc.load(product="ga_ls8c_gm_2_annual",
                         x=(24.60, 24.80),
                         y=(-20.05, -20.25),
                         time=("2018-01-01", "2018-12-31"),
                         output_crs="EPSG:32734",
                         resolution=(-250, 250))

print(ds_reprojected)

<xarray.Dataset>
Dimensions:  (time: 1, x: 87, y: 91)
Coordinates:
  * time     (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y        (y) float64 7.779e+06 7.779e+06 7.778e+06 ... 7.757e+06 7.756e+06
  * x        (x) float64 8.761e+05 8.764e+05 8.766e+05 ... 8.974e+05 8.976e+05
Data variables:
    red      (time, y, x) uint16 11483 11602 11831 11700 ... 11712 11785 11996
    green    (time, y, x) uint16 10618 10696 10935 10774 ... 10888 10949 11106
    blue     (time, y, x) uint16 9640 9667 9811 9771 ... 9757 9757 9808 9937
    nir      (time, y, x) uint16 15073 15243 15741 15275 ... 16055 16108 16468
    swir_1   (time, y, x) uint16 18086 18195 18438 18609 ... 17652 17711 18211
    swir_2   (time, y, x) uint16 15051 15104 15439 15649 ... 14384 14437 14869
Attributes:
    crs:      EPSG:32734


Note that the `crs` attribute in the *Attributes* section has changed to `EPSG:32734`. 
Due to the larger 250 m resolution, there are also now less pixels on the `x` and `y` dimensions (e.g. `x: 86, y: 91` compared to `x: 644, y: 801` in earlier examples).


### Spatial resampling methods
When a product is re-projected to a different CRS and/or resolution, the new pixel grid may differ from the original input pixels by size, number and alignment.
It is therefore necessary to apply a spatial "resampling" rule that allocates input pixel values into the new pixel grid.

By default, `dc.load()` resamples pixel values using "nearest neighbour" resampling, which allocates each new pixel with the value of the closest input pixel.
Depending on the type of data and the analysis being run, this may not be the most appropriate choice (e.g. for continuous data).

The `resampling` parameter in `dc.load()` allows you to choose a custom resampling method from the following options: 

```
"nearest", "cubic", "bilinear", "cubic_spline", "lanczos", 
"average", "mode", "gauss", "max", "min", "med", "q1", "q3"
```

For example, we can request that all loaded data is resampled using "average" resampling:

In [8]:
# Note the additional `resampling` parameter
ds_averageresampling = dc.load(product="ga_ls8c_gm_2_annual",
                                x=(24.60, 24.80),
                                y=(-20.05, -20.25),
                                time=("2018-01-01", "2018-12-31"),
                                resolution=(-250, 250),
                                resampling="average")

print(ds_averageresampling)


<xarray.Dataset>
Dimensions:  (time: 1, x: 78, y: 96)
Coordinates:
  * time     (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y        (y) float64 -2.507e+06 -2.508e+06 ... -2.531e+06 -2.531e+06
  * x        (x) float64 2.374e+06 2.374e+06 2.374e+06 ... 2.393e+06 2.393e+06
Data variables:
    red      (time, y, x) uint16 11854 11775 11553 11703 ... 11807 11749 11731
    green    (time, y, x) uint16 10940 10864 10653 10746 ... 10949 10900 10894
    blue     (time, y, x) uint16 9879 9849 9602 9609 ... 9932 9868 9779 9761
    nir      (time, y, x) uint16 15717 15571 15513 15754 ... 16123 16044 16069
    swir_1   (time, y, x) uint16 18827 18816 18131 18113 ... 18146 17788 17675
    swir_2   (time, y, x) uint16 15890 15906 14902 14743 ... 14905 14462 14346
Attributes:
    crs:      EPSG:6933


You can also provide a Python dictionary to request a different sampling method for different measurements. 
This can be particularly useful when some measurements contain contain categorical data which require resampling methods such as "nearest" or "mode" that do not modify the input pixel values.

In the example below, we specify `resampling={"red": "nearest", "*": "average"}`, which will use "nearest" neighbour resampling for the `red` satellite band only. `"*": "average"` will apply "average" resampling for all other satellite bands:


In [9]:
ds_customresampling = dc.load(product="ga_ls8c_gm_2_annual",
                                    x=(24.60, 24.80),
                                    y=(-20.05, -20.25),
                                    time=("2018-01-01", "2018-12-31"),
                                    resolution=(-250, 250),
                                    resampling={"red": "nearest", "*": "average"})

print(ds_customresampling)

<xarray.Dataset>
Dimensions:  (time: 1, x: 78, y: 96)
Coordinates:
  * time     (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y        (y) float64 -2.507e+06 -2.508e+06 ... -2.531e+06 -2.531e+06
  * x        (x) float64 2.374e+06 2.374e+06 2.374e+06 ... 2.393e+06 2.393e+06
Data variables:
    red      (time, y, x) uint16 11831 12056 11700 11468 ... 11867 11726 11738
    green    (time, y, x) uint16 10940 10864 10653 10746 ... 10949 10900 10894
    blue     (time, y, x) uint16 9879 9849 9602 9609 ... 9932 9868 9779 9761
    nir      (time, y, x) uint16 15717 15571 15513 15754 ... 16123 16044 16069
    swir_1   (time, y, x) uint16 18827 18816 18131 18113 ... 18146 17788 17675
    swir_2   (time, y, x) uint16 15890 15906 14902 14743 ... 14905 14462 14346
Attributes:
    crs:      EPSG:6933


> **Note**: For more information about spatial resampling methods, see the [following guide](https://rasterio.readthedocs.io/en/stable/topics/resampling.html)

## Loading data using the query dictionary syntax
It is often useful to re-use a set of query parameters to load data from multiple products.
To achieve this, we can load data using the "query dictionary" syntax.
This involves placing the query parameters we used to load data above inside a Python dictionary object which we can re-use for multiple data loads:

In [10]:
query = {"x": (24.60, 24.80),
         "y": (-20.05, -20.25),
         "time": ("2018-01-01", "2018-12-31")}


We can then use this query dictionary object as an input to `dc.load()`. 

> The `**` syntax below is Python's "keyword argument unpacking" operator.
This operator takes the named query parameters listed in the dictionary we created (e.g. `"x": (153.3, 153.4)`), and "unpacks" them into the `dc.load()` function as new arguments. 
For more information about unpacking operators, refer to the [Python documentation](https://docs.python.org/3/tutorial/controlflow.html#unpacking-argument-lists)

In [11]:
ds = dc.load(product="ga_ls8c_gm_2_annual",
             **query)

print(ds)

<xarray.Dataset>
Dimensions:  (time: 1, x: 644, y: 801)
Coordinates:
  * time     (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y        (y) float64 -2.507e+06 -2.507e+06 ... -2.531e+06 -2.531e+06
  * x        (x) float64 2.374e+06 2.374e+06 2.374e+06 ... 2.393e+06 2.393e+06
Data variables:
    red      (time, y, x) uint16 11808 11667 11901 11246 ... 11944 11948 11830
    green    (time, y, x) uint16 10801 10747 11000 10543 ... 11075 11068 10990
    blue     (time, y, x) uint16 9639 9581 9765 9467 ... 9892 9916 9926 9869
    nir      (time, y, x) uint16 15688 15618 15904 15293 ... 16541 16543 16316
    swir_1   (time, y, x) uint16 17885 17613 17969 16992 ... 18147 18170 17955
    swir_2   (time, y, x) uint16 14762 14453 14678 13979 ... 14850 14875 14693
Attributes:
    crs:      EPSG:6933


Query dictionaries can contain any set of parameters that would usually be provided to `dc.load()`:

In [12]:
query = {"x": (24.60, 24.80),
         "y": (-20.05, -20.25),
         "time": ("2018-01-01", "2018-12-31"),
         "output_crs": "EPSG:32734",
         "resolution": (-250, 250)}

ds_ls8 = dc.load(product="ga_ls8c_gm_2_annual",
                 **query)

print(ds_ls8)


<xarray.Dataset>
Dimensions:  (time: 1, x: 87, y: 91)
Coordinates:
  * time     (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y        (y) float64 7.779e+06 7.779e+06 7.778e+06 ... 7.757e+06 7.756e+06
  * x        (x) float64 8.761e+05 8.764e+05 8.766e+05 ... 8.974e+05 8.976e+05
Data variables:
    red      (time, y, x) uint16 11483 11602 11831 11700 ... 11712 11785 11996
    green    (time, y, x) uint16 10618 10696 10935 10774 ... 10888 10949 11106
    blue     (time, y, x) uint16 9640 9667 9811 9771 ... 9757 9757 9808 9937
    nir      (time, y, x) uint16 15073 15243 15741 15275 ... 16055 16108 16468
    swir_1   (time, y, x) uint16 18086 18195 18438 18609 ... 17652 17711 18211
    swir_2   (time, y, x) uint16 15051 15104 15439 15649 ... 14384 14437 14869
Attributes:
    crs:      EPSG:32734


Now we have a reusable query, we can easily use it to load data from a different product.
For example, we can load Water Observations from Space Annual Summary data for the same extent, time, output CRS and resolution that we just loaded Landsat 8 Geomedian data for:

In [13]:
ds_wofs = dc.load(product="ga_ls8c_wofs_2_annual_summary",
                 **query)

print(ds_wofs)

<xarray.Dataset>
Dimensions:      (time: 1, x: 87, y: 91)
Coordinates:
  * time         (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y            (y) float64 7.779e+06 7.779e+06 ... 7.757e+06 7.756e+06
  * x            (x) float64 8.761e+05 8.764e+05 ... 8.974e+05 8.976e+05
Data variables:
    count_wet    (time, y, x) int16 0 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0 0 0
    count_clear  (time, y, x) int16 35 36 36 36 36 36 36 ... 16 17 15 15 15 16
    frequency    (time, y, x) float32 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0
Attributes:
    crs:      EPSG:32734


## Other helpful tricks
### Loading data "like" another dataset
Another option for loading matching data from multiple products is to use `dc.load()`'s `like` parameter.
This will copy the spatial and temporal extent and the CRS/resolution from an existing dataset, and use these parameters to load a new data from a new product.

In the example below, we load another WOfS dataset that exactly matches the `ds_ls8` dataset we loaded earlier:


In [14]:
ds_wofs = dc.load(product="ga_ls8c_wofs_2_annual_summary",
                 like=ds_ls8)

print(ds_wofs)

<xarray.Dataset>
Dimensions:      (time: 1, x: 87, y: 91)
Coordinates:
  * time         (time) datetime64[ns] 2018-07-02T11:59:59.999500
  * y            (y) float64 7.779e+06 7.779e+06 ... 7.757e+06 7.756e+06
  * x            (x) float64 8.761e+05 8.764e+05 ... 8.974e+05 8.976e+05
Data variables:
    count_wet    (time, y, x) int16 0 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0 0 0
    count_clear  (time, y, x) int16 35 36 36 36 36 36 36 ... 16 17 15 15 15 16
    frequency    (time, y, x) float32 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0
Attributes:
    crs:      EPSG:32734


### Adding a progress bar
When loading large amounts of data, it can be useful to view the progress of the data load. 
The `progress_cbk` parameter in `dc.load()` allows us to add a progress bar which will indicate how the load is progressing. In this example, we will load 5 years of data (2013, 2014, 2015, 2016 and 2017) from the `ga_ls8c_wofs_2_annual_summary` product with a progress bar:

In [15]:
query = {"x": (24.60, 24.80),
         "y": (-20.05, -20.25),
         "time": ("2013", "2017")}

ds_progress = dc.load(product="ga_ls8c_wofs_2_annual_summary",
                      progress_cbk=with_ui_cbk(),
                      **query)

print(ds_progress)

<xarray.Dataset>
Dimensions:      (time: 5, x: 644, y: 801)
Coordinates:
  * time         (time) datetime64[ns] 2013-07-02T11:59:59.999500 ... 2017-07-02T11:59:59.999500
  * y            (y) float64 -2.507e+06 -2.507e+06 ... -2.531e+06 -2.531e+06
  * x            (x) float64 2.374e+06 2.374e+06 ... 2.393e+06 2.393e+06
Data variables:
    count_wet    (time, y, x) int16 0 0 0 0 0 0 0 0 0 0 ... 0 0 0 0 0 0 0 0 0 0
    count_clear  (time, y, x) int16 22 22 22 22 22 22 22 ... 17 18 18 18 18 18
    frequency    (time, y, x) float32 0.0 0.0 0.0 0.0 0.0 ... 0.0 0.0 0.0 0.0
Attributes:
    crs:      EPSG:6933


## Recommended next steps

For more advanced information about working with Jupyter Notebooks or JupyterLab, you can explore [JupyterLab documentation page](https://jupyterlab.readthedocs.io/en/stable/user/notebook.html).

To continue working through the notebooks in this beginner's guide, the following notebooks are designed to be worked through in the following order:

1. [Jupyter Notebooks](01_Jupyter_notebooks.ipynb)
2. [Products and Measurements](02_Products_and_measurements.ipynb)
3. **Loading data (this notebook)**
4. [Plotting](04_Plotting.ipynb)
5. [Performing a basic analysis](05_Basic_analysis.ipynb)
6. [Parallel processing with Dask](06_Parallel_processing_with_dask.ipynb)

Once you have you have completed the above six tutorials, join advanced users in exploring:

* The "Datasets" directory in the repository, where you can explore DE Africa products in depth.
* The "Frequently used code" directory, which contains a recipe book of common techniques and methods for analysing DE Africa data.
* The "Real-world examples" directory, which provides more complex workflows and analysis case studies.

***
## Additional information

**License:** The code in this notebook is licensed under the [Apache License, Version 2.0](https://www.apache.org/licenses/LICENSE-2.0). 

**Contact:** If you need assistance, please post a question on the [Open Data Cube Slack channel](http://slack.opendatacube.org/) or on the [GIS Stack Exchange](https://gis.stackexchange.com/questions/ask?tags=open-data-cube) using the `open-data-cube` tag (you can view previously asked questions [here](https://gis.stackexchange.com/questions/tagged/open-data-cube)).
If you would like to report an issue with this notebook, you can file one on [Github](https://github.com/GeoscienceAustralia/dea-notebooks).

**Last modified:** May 2020

**Compatible datacube version:** 

In [ ]:
print(datacube.__version__)

## Tags
Browse all available tags on the DE Africa User Guide's [Tags Index](https://) (placeholder as this does not exist yet)